[13.0] Split shopfloor core -> shopfloor_base

simahawk opened the pull request:
See atomic commits for details. Key changes:

* `shopfloor` base stuff moved to `shopfloor_base`
* `shopfloor` keeps all scenario and stock related features
* `schema` is an action component, not a service component anymore (in line w/ `data`)
* `scan_anything` made pluggable with components (eg: you can search what you want)
* new model `shopfloor.scenario` add to hold per-scenario configuration via YAML (special flags on s.menu are still there in `shopfloor`)
* shopfloor menu moved to root menu as an app (a specific menu entry allows to see logs from shopfloor app only)
* no change yet in `shopfloor_mobile`

**TODO:**

- [x] Update README
- [x] Change license to LGPL on shopfloor_base
- [x] Add missing TODO to ROADMAP

Next: split shopfloor_mobile /  shopfloor_mobile_base.

simahawk committed:
Add shopfloor_base (split from shopfloor)

simahawk committed:
shopfloor_base: add shopfloor.scenario

simahawk committed:
shopfloor_base: decouple actions lookup from service

simahawk committed:
shopfloor_base: 'schema' is an action as well

simahawk committed:
shopfloor_base: scan anything delegate search

Search and information retrieval is not hardcoded anymore.
You can register specific handlers from outside
to extend its behavior.

simahawk committed:
shopfloor_base: data|schema_detail are actions as well

simahawk committed:
shopfloor_base: actions_for must be private

The new api will try to load info for public methods on services.
Prevent it as having them public seems not relevant.

simahawk committed:
shopfloor_base: misc fixups

simahawk committed:
shopfloor_base: reorganize misc test classes

simahawk committed:
shopfloor_base: rework scan_anything tests

simahawk committed:
shopfloor: refactor on top of shopfloor_base

simahawk committed:
shopfloor: refactor scan_anything w/ components

simahawk committed:
shopfloor_batch_automatic_creation: refactor on shopfloor_base

simahawk committed:
shopfloor_checkout_sync: adapt to shopfloor_base

simahawk committed:
shopfloor: remove harcoded conf for scenario

Options can now be configured on scenario
hence is useless to have hardcoded behavior on menu model.

simahawk committed:
shopfloor_base: scan_anything allow limit record types scan

simahawk committed:
shopfloor_mobile: fix menu

simahawk committed:
shopfloor: checkout disable_no_package compat w/ yaml options

simahawk committed:
shopfloor_mobile: depend on shopfloor_base

simahawk committed:
shopfloor: fix rebase on 13.0

simahawk committed:
shopfloor: fix copyright

simahawk committed:
shopfloor_base: keep relation menu <-> scenario for existing records

simahawk committed:
shopfloor_packing_info: fix menu parent

simahawk committed:
shopfloor_workstation: fix menu parent

simahawk committed:
shopfloor_base: compute scenario for inactive menu

simahawk committed:
shopfloor_base: move menu to app, add quick link to logs

simahawk commented:
@lmignon @sebalix @hparfr here's the split. If possible, if no blocking change is detected, I'd like to speed up this merge to avoid painful rebase/cherry-pick/merge as we are still pushing fix/imp to the app. Also, this would allow to battle test this change in our customer project in real life and to speed up the split of the mobile module.

I've tested some scenario and everything seems fine (even if nothing has changed that much from the functional POV) but it's always better to see it live in action.

If you'll have improvement proposals I'll track them down and consider them for the next round.

simahawk committed:
shopfloor_workstation: fix demo data/test

simahawk committed:
shopfloor: fix after rebase on 13.0

lmignon commented on shopfloor_base/README.rst:
```diff
@@ -0,0 +1,193 @@
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+Shopfloor is a barcode scanner application for internal warehouse operations.
```
This description should be updated :smirk:

lmignon commented on shopfloor_base/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,38 @@
+Profiles
+~~~~~~~~
+
+In Inventory / Configuration / Shopfloor / Profiles.
```
The base shopfloor_base no more depends on `stock`.  `Inventory` as root into the path still relevant?

simahawk replied:
yeah sorry, I forgot to mention that I still have to cleanup READMEs :sweat_smile:

hparfr commented on shopfloor_base/models/shopfloor_scenario.py:
```diff
@@ -0,0 +1,70 @@
+    name = fields.Char(required=True, translate=True)
+    key = fields.Char(required=True)
+    options = Serialized(compute="_compute_options", default={})
+    options_edit = fields.Char(help="Configure options via YAML")
```
yaml ? 
Why not a sparse field like in storage ?

simahawk replied:
The sparse field is computed on it BUT this gives you much more flexibility in adding tech configurations w/out relying or real fields.
I find it very handy.

hparfr replied:
Introducing a whole lib just for that, it's a bit overkill no ?

It's less hype but JSON does job.

simahawk replied:
JSON sucks for manual editing. Adding pyyaml is not overkilling IMO.
It's an easy way to provide an easy way to edit options w/out getting insane w/ your own parser or hacks to display the values in a nice way (eg: body/body_str in se.index.config).

hparfr replied:
From an IT perspective : less dependencies = less conflicts + easier install

YAML has it's drawbacks too. It's still an hacky way to set things up, it's fine for the moment and I don't block it's addition even if I think it's not the good place to put it.

One more question: what are the extra settings you want to set?

simahawk replied:
Extra settings can be whatever you want. They are translated into `self.options` in [services](https://github.com/OCA/wms/blob/13.0/shopfloor/services/service.py#L284) and they can come from the work ctx as well.
You see an example here on how it can be used https://github.com/OCA/wms/blob/13.0/shopfloor/services/checkout.py#L79
Here you see the config for existing scenario https://github.com/simahawk/wms/blob/shopfloor-split/shopfloor/data/shopfloor_scenario_data.xml#L5
The point is that these options should be "dynamic" and ppl should be able to use them as they prefer, even with nested values or list of values, etc.

That been said, I'll remove YAML and rely on good writing by configurators :)

hparfr commented on shopfloor_base/models/shopfloor_scenario.py:
```diff
@@ -0,0 +1,70 @@
+    _description = "Shopfloor Scenario"
+
+    name = fields.Char(required=True, translate=True)
+    key = fields.Char(required=True)
```
what is the purpose of key ? 
Not blocking but, can't we work without it ?

simahawk replied:
it is mandatory: you really want 1 and only one scenario to match the usage of a component.

hparfr replied:
Why not a db id then ?

simahawk replied:
IDG your point. How could you do it via ID? This is something the integrator/dev should provide and it should match the usage on both back and front. It's exactly what was used before in `scenario` as string. 
I could change it to `usage` if you prefer the name but we need something like this anyway.

hparfr replied:
I suggested id because I misunderstood the meaning of key.

Does it replace that : https://github.com/OCA/wms/pull/169/files/ebce44f5d58c96bc8cc208c3f3e87d2f32d0abcd#diff-6b10fc808919519f099c8197a6ee5aadca6444fb042593847fe1066d23228657L84 ?

simahawk replied:
Yes, it replaces the string `scenario` that was on the menu model.

hparfr replied:
Personally, I prefer when this kind of fields are field Selection. From an UX perspective, I don't see why a user has to enter it itself. 

Because it's really a minor point, it's ok to leave it as is for the moment.

simahawk replied:
well, in theory the user is not the one that should edit this. I don't expect anyone to create scenario via UI as it would be useless. So, on the next round we should probably:

* make it not editable if not w/ very high permission
* make it a select

lmignon commented on shopfloor_base/__manifest__.py:
```diff
@@ -0,0 +1,41 @@
+    "depends": [
+        "base_jsonify",
+        "base_rest",
+        "rest_log",
```
Why do we have a dependency on `rest_log`? The only motivation seems to put some entries into the odoo menu :thinking:

simahawk replied:
well, is not required indeed. Maybe for now I can move it to `shopfloor` as we expect it to be there. Other long term option is having a dumb module `shopfloor_rest_log` but it seems too much for just having a menu item.

lmignon reviewed (commented):
At first read LGTM, some comments into the code.

hparfr commented on shopfloor_base/services/service.py:
```diff
@@ -0,0 +1,267 @@
+    def _get_openapi_default_parameters(self):
+        defaults = super()._get_openapi_default_parameters()
+        # Normal users can't read an API key, ignore it using sudo() only
+        # because it's a demo key.
```
IMO This code should be put somewhere else but not in the base module.

simahawk replied:
These params are mandatory for the whole app ATM. I won't change in this PR. I can add a TODO to discuss about it later.

hparfr replied:
ok

hparfr commented on shopfloor_base/services/service.py:
```diff
@@ -0,0 +1,267 @@
+    def msg_store(self):
+        return self._actions_for("message")
+
+    # TODO: maybe to be proposed to base_rest
```
May be it's time ?

simahawk replied:
well, IDK, Current solution is a bit hack-ish and w/ the new api we might be able to add this validation via decorators.
I'd keep it as it is now and get back to it on a second round.

hparfr replied:
Ok

hparfr reviewed (commented):
Just a code read for the moment.
Some minor remarks. Thanks for this great work.

simahawk committed:
shopfloor_base: pure json for scenario options edit

simahawk committed:
shopfloor: update README

simahawk committed:
shopfloor_base: update README

simahawk committed:
Add shopfloor_rest_log

simahawk committed:
shopfloor_base: drop dependency on rest_log

simahawk committed:
shopfloor_base: re-license to LGPL

simahawk commented:
@lmignon @hparfr done:

* update readme
* add shopfloor_rest_log (LGPL) and dropped dep on rest_log
* use json instead of yaml for options edit
* RE-LICENSE shopfloor_base TO LGPL
* rebase on 13.0 (all latest changes included)

simahawk commented:
rest_log license fix is here https://github.com/OCA/rest-framework/pull/114

simahawk commented:
@lmignon @hparfr are we happy here? :)

hparfr commented on shopfloor/models/shopfloor_menu.py:
```diff
@@ -2,51 +2,13 @@
-    profile_id = fields.Many2one(
-        "shopfloor.profile", string="Profile", help="Visible on this profile only"
-    )
     picking_type_ids = fields.Many2many(
```
`picking_type_ids` in not in the menu form view

simahawk replied:
ahh, good catch thanks!

simahawk replied:
fixed all missing fields / view override

hparfr reviewed (approved):
Code and functional review done.

Thank @simahawk for this big refactoring.

simahawk committed:
shopfloor_base: fix group menu by scenario

simahawk committed:
shopfloor_base: add TODO for scenario model

simahawk committed:
shopfloor: fix missing menu view override

simahawk committed:
shopfloor: fix menu.active field already in base

simahawk commented:
> Code and functional review done.
> 
> Thank @simahawk for this big refactoring.

thanks for your review! :wink:

simahawk commented:
I've already changed major version.

/ocabot merge minor

simahawk commented:
/ocabot merge minor

simahawk commented:
test failing and one possible permission issue: I'll check in a while

simahawk committed:
shopfloor_base: log what done via hooks

simahawk committed:
shopfloor: grant permission to stock users on update

simahawk commented:
bah, there's a test failing here https://travis-ci.com/github/OCA/wms/jobs/484254095#L1972 but I cannot reproduce it.
Likely something has changed somewhere in another PR or module or repo because the build on master is green but other PRs that do not contain the same changes as this one are failing for the same error. (eg: https://github.com/OCA/wms/pull/172)

simahawk commented:
Test fail because of this change https://github.com/odoo/odoo/commit/b33e72d0bf027fb2c789b1b9476f7edf1a40b0a6

simahawk committed:
shopfloor: fix package level unlink

https://github.com/odoo/odoo/commit/b33e72d0bf027fb2c789b1b9476f7edf1a40b0a6
changed the behavior of package level deletion on move lines.

simahawk committed:
shopfloor: improve _fill_stock_for_moves in tests

simahawk committed:
shopfloor: fix stock.move.extract_and_action_done

There's no need to call action assign on the brand new backorder
because all lines at that point will be already assigned.

simahawk commented:
last 3 commits fixes compatibilty w/ odoo core change https://github.com/odoo/odoo/commit/b33e72d0bf027fb2c789b1b9476f7edf1a40b0a6

- shopfloor: fix stock.move.extract_and_action_done
- shopfloor: improve _fill_stock_for_moves in tests
- shopfloor: fix package level unlink

simahawk commented:
/ocabot merge nobump

simahawk commented:
Similar PR for mobile https://github.com/OCA/wms/pull/175
